### PR TITLE
Add additional documentation and examples to `ArrayAccessor`

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

N/A

# Rationale for this change
 
While reviewing https://github.com/apache/datafusion/pull/11676 from @XiangpengHao I wanted to try and increase the liklihood that others can write such generic code. 

I thought the example from that  code from that PR into an example

# What changes are included in this PR?
Improve doc comments, add example in `ArrayAccessor`

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-array/src/array/mod.rs [hunk 1]
```diff
@@ -437,13 +437,84 @@ impl<'a, T: Array> Array for &'a T {
 
 /// A generic trait for accessing the values of an [`Array`]
 ///
+/// This trait helps write specialized implementations of algorithms for
+/// different array types. Specialized implementations allow the compiler
+/// to optimize the code for the specific array type, which can lead to
+/// significant performance improvements.
+///
+/// # Example
```
Comment (alamb) [inline arrow-array/src/array/mod.rs:445]: This example is adapted from https://github.com/apache/datafusion/pull/11676

File: arrow/src/lib.rs [hunk 2]
```diff
@@ -81,10 +86,11 @@
 //!
 //! # Type Erasure / Trait Objects
 //!
-//! It is often the case that code wishes to handle any type of array, without necessarily knowing
-//! its concrete type. This use-case is catered for by a combination of [`Array`]
-//! and [`DataType`](datatypes::DataType), with the former providing a type-erased container for
-//! the array, and the latter identifying the concrete type of array.
+//! It is common to write code that handles any type of array, without necessarily
```
Comment (alamb) [inline arrow/src/lib.rs:89]: I tried to make this less verbose so it was easier to read.

**Comment by XiangpengHao** [comment]
Some closely related code: https://github.com/apache/arrow-rs/blob/705d3414eb991c234aaced332c11973e997d360f/arrow-string/src/like.rs#L151-L161

**Review by tustvold — APPROVED** [review]

**Comment by alamb** [comment]
Thank you for the review @tustvold

**Pull request merged**